### PR TITLE
DEBUG0 bug fix, more chas and services supported

### PR DESCRIPTION
**Pull request by Andreaqc**
If DEBUG0 is not defined there is a syntax error.
More chas and more services defined and supported.

### CODE REVIEW

**Comment by HomeACcessoryKid** [comment]
Hi Andrea,

Many thanks for your contributions.
I would like to ask if you tested (real good) the changes you made to e.g. the decrypt function, the MAXITM etc.?? I do not only mean functionality but also memory consumption, an important issue in an ESP8266?  
I recall that I made those lines off code the way they are on purpose and after extensive testing. And debugging it later will be a pain.

Second, I have not much experience with GIThub so I want to be really careful of what I do. And I am very busy in real life as well.
All this means that I will take more time than I would like to.
Finally, I cannot accept the change in password, it is documented to be the original value and will confuse everybody.

Thanks again for your help and I also invite you to merge your Sonoff 'Demo' to the Demo repository, nice work!

**Comment by Andreaqc** [comment]
Hello,

i am actually using a sonoff with my fixes on and it is working flawlessly.

I increased MAXITM to support more services at once on a single device (Thermostat and a fan on a single esp for example) and i saw that memory is pretty much stable, the only problem could be cJSON_PrintUnformatted on a big json. Maybe it needs a bit more testing to find the sweet spot. 

A bigger URLSize is necessary for services with more than 4 characteristics otherwise we can't get the full request. However it is completely safe to increase it.

The decrypt function was wrong and it was causing a loss of "bigger-than-usual" packets usually sent after the pair-verify session.

I am sorry for the change in password, that was my fault, however i changed it back to default as you can see [here](https://github.com/HomeACcessoryKid/ESP8266-HomeKit/pull/29/commits/f98ed0237869797ed20e8bbf8d5ed762206884c5).

Thanks for sharing your code, you did great job!

File: hkc/hkc.c [hunk 1]
```diff
@@ -53,11 +53,11 @@
 #include "hk.h"
 
 #define NLEN    384
-#define MAXITM   15
+#define MAXITM   32
 
 extern  espconn_msg *plink_active;
 //below the global struct for the acc_items
-acc_item    acc_items[MAXITM+1];
```
Comment (HomeACcessoryKid) [inline hkc/hkc.c]: This +1 is there for a reason. considering Apple starts numbering items from 1, we need to keep the meaning intact and make an array with one item more, which is item 0 which we will never use. If now you would add exactly 32 items, it will coredump. Please undo this change

File: hkc/hkc.c [hunk 2]
```diff
@@ -181,7 +181,6 @@ char    *parse_cgi(char *in) //take aid.iid string and return chars string / onl
     char *out;
     cJSON *chars,*items,*item;
     int aid, iid;
-
```
Comment (HomeACcessoryKid) [inline hkc/hkc.c]: in general I prefer to keep an empty line between declarations an code. Please change back

File: hkc/hkc.c [hunk 3]
```diff
@@ -383,74 +383,163 @@ cJSON   *addService(cJSON *services, int iid, char *brand, int sType)
...
     switch (cType) {
+        case BRIGHTNESS_C: {
+            strcpy(format,INT);         perm=7;     maxlen=0;   ev=1;
+            cJSON_AddNumberToObject(acc_items[iid].json, "minValue",   0);
+            cJSON_AddNumberToObject(acc_items[iid].json, "maxValue", 100);
+            cJSON_AddNumberToObject(acc_items[iid].json, "minStep",    1);
+            cJSON_AddStringToObject(acc_items[iid].json, "unit", "percentage");
+        } break;
+        case CURRENT_HEATING_COOLING_STATE_C:{
+            strcpy(format,UINT8);        perm=5;
```
Comment (HomeACcessoryKid) [inline hkc/hkc.c]: i think the specs say that this item should also have ev=1;  
please verify if ALL the items you added conform to spec?

    {
      "Constraints": {
        "ValidValues": {
          "0": "Off",
          "1": "Heat",
          "2": "Cool"
        }
      },
      "Name": "Current Heating Cooling State",
      "UUID": "0000000F-0000-1000-8000-0026BB765291",
      "Properties": [
        "read",
        "cnotify",
        "uncnotify"
      ],
      "Format": "uint8",
      "Permissions": [
        "securedRead"
      ]
    },

File: hkc/hkc.c [hunk 4]
```diff
@@ -383,74 +383,163 @@ cJSON   *addService(cJSON *services, int iid, char *brand, int sType)
...
+        case CURRENT_TEMPERATURE_C:{
+            strcpy(format,FLOAT);        perm=5;
+            cJSON_AddNumberToObject(acc_items[iid].json, "minValue",   0);
+            cJSON_AddNumberToObject(acc_items[iid].json, "maxValue",   100);
+            cJSON_AddNumberToObject(acc_items[iid].json, "minStep",    0.1);
+            cJSON_AddStringToObject(acc_items[iid].json, "unit", "celsius");   
+        } break;
         case IDENTIFY_C: {
             strcpy(format,BOOLEAN);     perm=2;     maxlen=1;
-            cJSON_AddStringToObject(acc_items[iid].json, "description", "Identify");
```
Comment (HomeACcessoryKid) [inline hkc/hkc.c]: does it still work after?

File: hkc/hkc.c [hunk 5]
```diff
@@ -383,74 +383,163 @@ cJSON   *addService(cJSON *services, int iid, char *brand, int sType)
...
-            strcpy(format,STRING);      perm=4;     maxlen=255;
-            cJSON_AddStringToObject(acc_items[iid].json, "description", "Serial");
+            strcpy(format,STRING);      perm=4;     maxlen=64;
         } break;
         case NAME_C: {
-            strcpy(format,STRING);      perm=4;     maxlen=255;
-            cJSON_AddStringToObject(acc_items[iid].json, "description", "Name");
+            strcpy(format,STRING);      perm=4;     maxlen=64;
         } break;
-        case POWER_STATE_C: {
```
Comment (HomeACcessoryKid) [inline hkc/hkc.c]: this is needed for backward compatibility with the Demo code... it can still be the same switch case

File: hkc/hkc.c [hunk 6]
```diff
@@ -383,74 +383,163 @@ cJSON   *addService(cJSON *services, int iid, char *brand, int sType)
...
+            cJSON_AddItemToObject(acc_items[iid].json, "valid-values", valid_values=cJSON_CreateArray());  
+            cJSON_AddItemToArray(valid_values,cJSON_CreateNumber(0));       
+            cJSON_AddItemToArray(valid_values,cJSON_CreateNumber(1)); 
         } break;
         default: {
             
         } break;
     }
-    if (ev) cJSON_AddTrueToObject( acc_items[iid].json, "events");
-    else    cJSON_AddFalseToObject(acc_items[iid].json, "events");
```
Comment (HomeACcessoryKid) [inline hkc/hkc.c]: The concept of event is a feature... why did you remove it? does it 'happen' automatically based on perms="ev"?
But the spec says 'events' is an attribute by its own right so we must set it in my view

File: hkc/hkc.c [hunk 7]
```diff
@@ -383,74 +383,163 @@ cJSON   *addService(cJSON *services, int iid, char *brand, int sType)
...
     //encode perms like rwe octal
     if (perm & 2) cJSON_AddItemToArray(perms,cJSON_CreateString("pw"));
     if (perm & 4) cJSON_AddItemToArray(perms,cJSON_CreateString("pr"));
     if (perm & 1) cJSON_AddItemToArray(perms,cJSON_CreateString("ev"));
     //addItem(aid,iid,format,valuestring,change_cb);
     if (valuestring) {
         if (!strcmp(format,BOOLEAN)){
-            if ( !strcmp(valuestring,"0") || !strcmp(valuestring,"false") ) intval=0; else intval=1;
-            cJSON_AddItemToObject(acc_items[iid].json, "value", value=cJSON_CreateBool(intval) );
+            intval = 0;
```
Comment (HomeACcessoryKid) [inline hkc/hkc.c]: no actual functional improvement, just a matter of taste please remove, so people do not spend time on wondering what has changed

File: hkc/hkc.c [hunk 8]
```diff
@@ -929,7 +1018,7 @@ server_recv(void *arg, char *pusrdata, unsigned short length)
 {
     struct espconn *ptrespconn = arg;
     crypto_parm *pcryp = ptrespconn->reserve;
-    if (pcryp && xSemaphoreTake(pcryp->semaphore,0)){
```
Comment (HomeACcessoryKid) [inline hkc/hkc.c]: Please explain why you changed this?

In my opinion this NEEDS to be a polling. we will ignore all requests from the same connection_memory_block if the previous one is still not released because operations on it are still ongoing. If not, we set up a new connection and later the callback of the old one will coredump the device.  
It took months to get everything stable (which now it is!)  
Read up on esp_conn before touching this again.  Suggestions/questioning are of course more than welcome.

File: hkc/hkc.c [hunk 9]
```diff
@@ -1684,8 +1773,8 @@ void crypto_init()
     #endif
     if (strcmp(flash,signature)) {
         #ifdef DEBUG0
-        os_printf("initializing flash in 15 seconds\n");
-        vTaskDelay(3000);
```
Comment (HomeACcessoryKid) [inline hkc/hkc.c]: This needs to be there to be able to catch the moment of initialising. My terminal program takes >2 second to set up after I finish flashing. It can also be a warning for people to power down and save the flash content. please revert. It is not affecting normal day usage.

File: hkc/hkc.c [hunk 10]
```diff
@@ -2454,7 +2543,7 @@ void decrypt(void *arg, char *data, unsigned short *length)  // length will chan
     buffer = (byte *)zalloc(*length);
     total=*length; *length=0;
     for (offset=0;offset<total;){
-        len = 255*data[1]+data[0]; //Little Endian
+        len = 256*data[1]+data[0]; //Little Endian
```
Comment (HomeACcessoryKid) [inline hkc/hkc.c]: it appears to be a valid fix, I guess the phone never sent anything substantial till now to trigger this bug...
good catch!

File: hkc/include/hkc.h [hunk 11]
```diff
@@ -111,53 +111,151 @@ void    addCharacteristic(cJSON *chas, int aid, int iid, char *brand, int cType,
...
-#define VERSION_C                               0x37
-#define CURRENTHEATINGCOOLING_C                 0x0F
-#define TARGETHEATINGCOOLING_C                  0x33
+//cType name                                        Type    //mxlen format  read/write/event
+#define ADMINISTRATOR_ONLY_ACCESS_C                 0x01
+#define AUDIO_FEEDBACK_C                            0x05
+#define BRIGHTNESS_C                                0x08    //n/a   int     rwe
+#define COOLING_THRESHOLD_TEMPERATURE_C             0x0D
+#define CURRENT_DOOR_STATE_C                        0x0E
+#define CURRENT_HEATING_COOLING_STATE_C             0x0F
```
Comment (HomeACcessoryKid) [inline hkc/include/hkc.h]: please add the mxlen, format and rwe to each of the items you have implemented. This is the face of the API and this feedback is useful for those that do not debug the foundation code.

File: hkc/include/hkc.h [hunk 12]
```diff
@@ -111,53 +111,151 @@ void    addCharacteristic(cJSON *chas, int aid, int iid, char *brand, int cType,
...
+#define LOCK_LAST_KNOWN_ACTION_C                    0x1C
+#define LOCK_MANAGEMENT_AUTO_SECURITY_TIMEOUT_C     0x1A
+#define LOCK_TARGET_STATE                           0x1E
+#define LOGS_C                                      0x1F
+#define MANUFACTURER_C                              0x20    //255   string  r
+#define MODEL_C                                     0x21    //255   string  r
+#define MOTION_DETECTED_C                           0x22
+#define NAME_C                                      0x23    //255   string  r
+#define OBSTRUCTION_DETECTED_C                      0x24
+#define ON_C                                        0x25    //  1   boolean rwe
```
Comment (HomeACcessoryKid) [inline hkc/include/hkc.h]: add the POWER_STATE item and comment behind that it is for backward compatibility

File: hkc/hkc.c (hunk 3 above)
Comment (Andreaqc) [reply hkc/hkc.c]: Ev is now inferred py perm.

File: hkc/hkc.c [hunk 13]
```diff
@@ -383,74 +383,162 @@ cJSON   *addService(cJSON *services, int iid, char *brand, int sType)
...
-    if (perm & 1) cJSON_AddItemToArray(perms,cJSON_CreateString("ev"));
+    if (perm & 1) {
+        cJSON_AddItemToArray(perms, cJSON_CreateString("ev"));
+        cJSON_AddTrueToObject( acc_items[iid].json, "events");
+    }
     //addItem(aid,iid,format,valuestring,change_cb);
     if (valuestring) {
         if (!strcmp(format,BOOLEAN)){
-            if ( !strcmp(valuestring,"0") || !strcmp(valuestring,"false") ) intval=0; else intval=1;
+            if ( !strcmp(valuestring,"0") || !strcmp(valuestring,"false") ) intval=0; else intval=1;
```
Comment (HomeACcessoryKid) [inline hkc/hkc.c]: please remove the gremlin

File: hkc/hkc.c [hunk 14]
```diff
@@ -383,74 +383,162 @@ cJSON   *addService(cJSON *services, int iid, char *brand, int sType)
...
+        if(!strcmp(format,STRING))
+            cJSON_AddNumberToObject(acc_items[iid].json, "maxLen", maxlen );
+        else if (!strcmp(format,DATA))
+            cJSON_AddNumberToObject(acc_items[iid].json, "maxDataLen", maxlen );
+    }
     //encode perms like rwe octal
     if (perm & 2) cJSON_AddItemToArray(perms,cJSON_CreateString("pw"));
     if (perm & 4) cJSON_AddItemToArray(perms,cJSON_CreateString("pr"));
-    if (perm & 1) cJSON_AddItemToArray(perms,cJSON_CreateString("ev"));
+    if (perm & 1) {
```
Comment (HomeACcessoryKid) [inline hkc/hkc.c]: please keep the original line for perm&1 and fix 'events' at the original spot

File: hkc/hkc.c (hunk 6 above)
Comment (HomeACcessoryKid) [reply hkc/hkc.c]: please replace ev with perm&1 (potentially move this construct below the other perm& units

File: hkc/hkc.c (hunk 2 above)
Comment (HomeACcessoryKid) [reply hkc/hkc.c]: please remove the gremlin

File: hkc/hkc.c [hunk 15]
```diff
@@ -383,74 +383,162 @@ cJSON   *addService(cJSON *services, int iid, char *brand, int sType)
...
         } break;
         case MODEL_C: {
-            strcpy(format,STRING);      perm=4;     maxlen=255;
-            cJSON_AddStringToObject(acc_items[iid].json, "description", "Model");
+            strcpy(format,STRING);      perm=4;     maxlen=64;
+        } break;
+        case NAME_C: {
+            strcpy(format,STRING);      perm=4;     maxlen=64;
+        } break;
+        case ON_C: {
```
Comment (HomeACcessoryKid) [inline hkc/hkc.c]: power state was expected here...

File: hkc/hkc.c (hunk 14 above)
Comment (HomeACcessoryKid) [reply hkc/hkc.c]: maybe you are right in not encoding events=false
in that case add one line next to the original like this:
`
    if (perm & 1) cJSON_AddItemToArray(perms,cJSON_CreateString("ev"));
    if (perm & 1) cJSON_AddTrueToObject( acc_items[iid].json, "events");
`

Comment (Andreaqc) [reply hkc/hkc.c]: There is no point in doing this.

File: hkc/hkc.c (hunk 15 above)
Comment (Andreaqc) [reply hkc/hkc.c]: Please stick to the official names, "Power State" is not the official name of this characteristic, i suggest to update your demo code.
However i defined POWER_STATE_C for backward compatibility in the headers.

File: hkc/include/hkc.h [hunk 16]
```diff
@@ -111,53 +111,152 @@ void    addCharacteristic(cJSON *chas, int aid, int iid, char *brand, int cType,
...
+#define HEATING_THRESHOLD_TEMPERATURE_C             0x12
+#define HUE_C                                       0x13
+#define IDENTIFY_C                                  0x14    //n/a   boolean w
+#define LOCK_CONTROL_POINT_C                        0x19
+#define LOCK_CURRENT_STATE_C                        0x1D
+#define LOCK_LAST_KNOWN_ACTION_C                    0x1C
+#define LOCK_MANAGEMENT_AUTO_SECURITY_TIMEOUT_C     0x1A
+#define LOCK_TARGET_STATE                           0x1E
+#define LOGS_C                                      0x1F
+#define MANUFACTURER_C                              0x20    //255   string  r
```
Comment (HomeACcessoryKid) [inline hkc/include/hkc.h]: the maxlen values have been replaced by 64 so should be here as well

File: hkc/include/hkc.h [hunk 17]
```diff
@@ -111,53 +111,152 @@ void    addCharacteristic(cJSON *chas, int aid, int iid, char *brand, int cType,
...
+#define MODEL_C                                     0x21    //255   string  r
+#define MOTION_DETECTED_C                           0x22
+#define NAME_C                                      0x23    //255   string  r
+#define OBSTRUCTION_DETECTED_C                      0x24
+#define ON_C                                        0x25    //n/a   boolean rwe
+#define POWER_STATE_C                               0x25    //n/a   boolean rwe
```
Comment (HomeACcessoryKid) [inline hkc/include/hkc.h]: comment should read: //deprecated for ON_C

File: hkc/hkc.c (hunk 15 above)
Comment (HomeACcessoryKid) [reply hkc/hkc.c]: you got a point, header definition is enough
Demo code will be updated as well

File: hkc/include/hkc.h [hunk 18]
```diff
@@ -111,53 +111,152 @@ void    addCharacteristic(cJSON *chas, int aid, int iid, char *brand, int cType,
...
+#define TARGET_VERTICAL_TILT_ANGLE_C                0x7D
+#define SECURITY_SYSTEM_ALARM_TYPE_C                0x8E
+#define CHARGING_STATE_C                            0x8F
+#define CARBON_MONOXIDE_LEVEL_C                     0x90
+#define CARBON_MONOXIDE_PEAK_LEVEL_C                0x91
+#define CARBON_DIOXIDE_DETECTED_C                   0x92
+#define CARBON_DIOXIDE_LEVEL_C                      0x93
+#define CARBON_DIOXIDE_PEAK_LEVEL_C                 0x94
+#define AIR_QUALITY_C                               0x95
+#define STREAMING_STATUS_C                          0x120
```
Comment (HomeACcessoryKid) [inline hkc/include/hkc.h]: the 3 digits should be below the 2 digits for consistency

**Comment by HomeACcessoryKid** [comment]
final comment, is it possible to put fixes of different categories in different pull requests? Now it was all one big ball with a all-or-nothing approval.

Anyway, sorry for the delay and thanks for your work.
